src/sources/connect/query_graph.rs

```diff
@@ -0,0 +1,812 @@
+};
+
+use super::{
+    selection_parser::{NamedSelection, PathSelection, Property, SubSelection},
+    spec::schema::{ConnectDirectiveArguments, CONNECT_DIRECTIVE_NAME_IN_SPEC},
+    ConnectFederatedConcreteFieldQueryGraphEdge, ConnectFederatedConcreteQueryGraphNode,
+    ConnectFederatedEnumQueryGraphNode, ConnectFederatedQueryGraphBuilder, Selection,
+};
+
+impl SourceFederatedQueryGraphBuilderApi for ConnectFederatedQueryGraphBuilder {
+    fn process_subgraph_schema(
+        &self,
+        subgraph: ValidFederationSubgraph,
+        builder: &mut impl IntraSourceQueryGraphBuilderApi,
+    ) -> Result<Vec<FederatedLookupTailData>, FederationError> {
+        // Grab all connects
+        // TODO: Handle renames
+        let connect_references = subgraph
+            .schema
+            .referencers()
+            .get_directive(&CONNECT_DIRECTIVE_NAME_IN_SPEC)?;
+
+        // Extract the connects from the schema definition and map them to their `Connect` equivalent
+        // Note: We need to keep the object_field as well as the arguments since it contains important
+        // type information of the original schema.
+        // TODO: Validate that we don't have connect directives on anything but object fields.
+        let connects: Vec<_> = connect_references
+            .object_fields
+            .iter()
+            .map(|object_field| {
+                // Note: We need to keep the `object_field` around since it contains metadata needed for graph
+                // creation.
+                object_field
+                    .get(subgraph.schema.schema())
+                    .map(move |f| (object_field, f))
+            })
+            .map_ok(|(object_field, field)| {
+                field
+                    .directives
+                    .iter()
+                    .map(move |d| (object_field, field, d))
+            })
+            .flatten_ok()
+            .map_ok(|(object_field, field, directive)| {
+                ConnectDirectiveArguments::try_from(directive)
+                    .map(move |args| (object_field, field, args))
+            })
+            .flatten_ok()
+            .try_collect()?;
+
+        for (object, field, args) in connects {
+            // Make a node for the entrypoint of this field, if not yet created
+            let parent_node = builder.add_concrete_node(
+                object.type_name.clone(),
+                SourceFederatedConcreteQueryGraphNode::Connect(
+                    ConnectFederatedConcreteQueryGraphNode::ConnectParent {
+                        subgraph_type: object.parent().clone(),
+                    },
+                ),
+            )?;
+
+            // Process the field, constructing the rest of the graph for its selections
+            // TODO: What should we do if the selection is empty? Make a selection with all fields?
+            let field_node =
+                process_selection(args.selection.unwrap(), field, &subgraph.schema, builder)?;
```

Comment:
if we're going to traverse the selection AST as-in, then we can delete the `From<Selection> for GraphQLSelection` code here https://github.com/apollographql/federation-next/blob/connectors/src/sources/connect/selection_parser.rs#L734

i think this makes sense because we'll want to keep track of the complete JSONSelection component with each edge (not just the "left side" of the selection that relates to the schema)